MSC4146: Shared Message Drafts

cyrneko opened the pull request:
[Rendered](https://github.com/cyrneko/matrix-spec-proposals/blob/shared-message-drafts/proposals/4146-Shared-Message-Drafts.md)

This MSC proposes a way to share message drafts between clients, to edit and/or send them on clients they weren't initially written in.

Signed off by: Simon Müller (simeonlps@proton.me)

cyrneko committed:
Add MSCXXXX-Shared-Message-Drafts.md

cyrneko committed:
Update and rename MSCXXXX-Shared-Message-Drafts.md to 4146-Shared-Message-Drafts.md

clokep commented on proposals/4146-Shared-Message-Drafts.md:
```diff
@@ -0,0 +1,99 @@
+{
+  "type": "m.room_drafts",
+  "content": {
+    "!ROOMID:example.com": {
```
Using [room specific account data](https://spec.matrix.org/v1.10/client-server-api/#get_matrixclientv3useruseridroomsroomidaccount_datatype) could remove a layer of mapping in this.

TGRush replied:
I'll look into that, thank you! :)

uhoreg replied:
Account data would have been my first guess for implementing this feature, but you'd need to figure out some encryption scheme to avoid leaking data to the server.

clokep commented on proposals/4146-Shared-Message-Drafts.md:
```diff
@@ -0,0 +1,99 @@
+| prefix                        | description                       |
+|-------------------------------|-----------------------------------|
+| `org.matrix.msc4146.drafts`   | For the room type                 |
+| `org.matrix.msc4146.draft_in` | For the actual events in the room |
```
I don't think this matches the proposal?

```suggestion
| `org.matrix.msc4146.draft_in` | For the room ID information on the events |
```

TGRush replied:
fair change, I see no issues

clokep commented on proposals/4146-Shared-Message-Drafts.md:
```diff
@@ -0,0 +1,99 @@
+
+This proposal aims to address that problem by sharing ideas on a mechanism to share message drafts between devices, and serves as a place of discussion for such a feature.
+
+This being my first MSC, without having read the entire spec there will be mistakes and oversights I've made as to how Matrix operates and how a feature like this may be implemented, so I'd very much appreciate some pointers or help if I got anything wrong :)
```
Thanks for providing this context! But I think it can be removed from the actual MSC text which is meant to be mostly technical. 😄

TGRush replied:
Got it! Thanks for the heads-up

clokep commented on proposals/4146-Shared-Message-Drafts.md:
```diff
@@ -0,0 +1,99 @@
+
+## Proposal
+
+In Matrix, especially when used in work environments, there is often the need to draft out a message before sending it out, like wanting someone else to proofread it first, writing a message about an ongoing event, or a plethora of other scenarios where a message should just remain unsent until the time comes. However the experience is frustrating because you are limited only to one device, or ev ...
+
+If you're ever in a situation where you want to send the message, edit it, or maybe even delete the draft, you cannot do so without first going back to the device you initially wrote it on. This means you cannot carry over a draft from your Phone to your Computer to have a more comfortable typing experience, or to see the text on a bigger screen, and is generally just an inconvenience.
```
This feels like it is mostly duplicating the intro, maybe combine it?

TGRush replied:
I did try a bit *too* hard on following the template maybe, but yeah that is (again) a fair point

clokep commented on proposals/4146-Shared-Message-Drafts.md:
```diff
@@ -0,0 +1,99 @@
+
+If you're ever in a situation where you want to send the message, edit it, or maybe even delete the draft, you cannot do so without first going back to the device you initially wrote it on. This means you cannot carry over a draft from your Phone to your Computer to have a more comfortable typing experience, or to see the text on a bigger screen, and is generally just an inconvenience.
+
+My first idea would be to repurpose rooms, similar to how it's proposed for Profiles-As-Rooms or how Spaces are implemented. A new room type, `m.drafts` or `m.draft` would be added, allowing the client to differentiate this room from DMs, Spaces or Groups. A creation event could look something like this:
```
Write this as if it is the change:

```suggestion
Rooms will be repurposed to store drafts, similar to how it's proposed for Profiles-As-Rooms or how Spaces are implemented. A new room type, `m.drafts` would be added to allow clients to differentiate it from a normal room. A creation event could look something like this:
```

TGRush replied:
Thank you!

clokep commented on proposals/4146-Shared-Message-Drafts.md:
````diff
@@ -0,0 +1,99 @@
+}
+```
+
+## Potential issues
````
Is it valid for other members to be in that room? What happens if a user is in multiple `m.drafts` rooms?

TGRush replied:
I think multiple members apart from yourself to be in the room would be fine, but I will have to clarify in the proposal that the client should only read it's own events (e.g don't import drafts from other users), unless explicitly requested otherwise by the user

As for being in multiple `m.drafts` rooms, the client should take whatever is chronologically the newest draft for any given room, unless there is two drafts for one room with the same timestamp. In that case the user should be given the option to pick between which draft to keep, and have the other draft redacted.

Thanks!

clokep commented on proposals/4146-Shared-Message-Drafts.md:
```diff
@@ -0,0 +1,99 @@
+
+## Potential issues
+
+*Not all proposals are perfect. Sometimes there's a known disadvantage to implementing the proposal,
+and they should be documented here. There should be some explanation for why the disadvantage is
+acceptable, however - just like in this example.*
```
I think this is from the template?

```suggestion
```

TGRush replied:
oh boy yeah, I forgot to remove that one :upside_down_face:

clokep commented on proposals/4146-Shared-Message-Drafts.md:
```diff
@@ -0,0 +1,99 @@
+and they should be documented here. There should be some explanation for why the disadvantage is
+acceptable, however - just like in this example.*
+
+There are some issues with both approaches, notably I thought of:
```
What do you mean by "both" here? What's the other approach?

TGRush replied:
That is a little leftover from when I was writing this locally, I initially had the structure of the document a bit differently so it made sense to write it that way, but after I thought of encryption I moved the 2nd idea into the *Alternatives* section instead.

I'll remove/fix this in the next commit

cyrneko committed:
Update proposals/4146-Shared-Message-Drafts.md

Co-authored-by: Patrick Cloke <clokep@users.noreply.github.com>

cyrneko committed:
Update 4146-Shared-Message-Drafts.md

cyrneko committed:
Update 4146-Shared-Message-Drafts.md

TGRush commented on proposals/4146-Shared-Message-Drafts.md:
````diff
@@ -0,0 +1,99 @@
+}
+```
+
+In said room, the client would then send normal `m.text` events, optionally with a formatted body, and would edit the message every ´n´ seconds or once the user stops typing. <span style="color: grey">(This is more of a UI/UX implementation detail though)</span>
````
I figured it would be easiest to either only save the plaintext or formatted body of an event because from what I've seen, the client would be parsing the input anyways before sending it off as an event

So if you are mentioning someone in a draft, and it gets synced to another device, the client would add intentional mentions anyways. I made this decision in an effort to minimize the amount of data that needs to be synced, but I think having a copy of the *whole* event as it would be sent out may actually be beneficial if you're drafting something with Media for example :thinking: 

I'll reconsider this for sure, thank you!

clokep replied:
Yeah, I think that makes sense. Just clarifying the idea here.

cyrneko committed:
Update proposals/4146-Shared-Message-Drafts.md

Co-authored-by: Patrick Cloke <clokep@users.noreply.github.com>

cyrneko committed:
Update 4146-Shared-Message-Drafts.md

cyrneko committed:
Update 4146-Shared-Message-Drafts.md

turt2live commented on proposals/4146-Shared-Message-Drafts.md:
```diff
@@ -0,0 +1,95 @@
+
+## Proposal
+
+There is often the need to draft out a message before sending it out, like wanting someone else to proofread it first, writing a message about an ongoing event, or a plethora of other scenarios where a message should just remain unsent until the time comes. However the experience is frustrating because you are limited only to one device, or even just one client on the same device. 
```
(please wrap lines to ~100 characters)

cyrneko replied:
oh sorry, I completely missed that! I'll get to it in a moment :)

cyrneko replied:
had something to do, it should be wordwrapped at ~80 characters now :)

turt2live commented on proposals/4146-Shared-Message-Drafts.md:
Implementation requirements:

* [ ] Client recognizing and using the feature

cyrneko committed:
wordwrap

cyrneko committed:
re-add MSC numbers 

:facepalm:

cyrneko committed:
I should triple-check my diffs....

oversight where I took a local copy and wordwrapped it when I already had changes done remotely :facepalm:

cyrneko committed:
media and `m.draft_full_event`

Added a section about drafting media (`m.image` in this example) and also how you could draft a whole event. This could be useful for polls, custom events, or drafting messages with other MSCs.

cyrneko committed:
unstable prefix for `m.draft_full_event`

uhoreg commented on proposals/4146-Shared-Message-Drafts.md:
```diff
@@ -0,0 +1,198 @@
+- If a client removes the `m.draft_in` as part
+of an edit the draft would no longer be associated with a room, causing
+confusion on the users' end as to where their draft went.
+
```
Another issue is that converting to and from a user's input to a Matrix message can be lossy.  For example, if a user starts writing a message on a client that uses Markdown, and resumes editing later on a different client that uses a different input format, some formatting may get lost.

cyrneko replied:
I think that shouldn't be *too* much of a problem considering the plaintext should (in theory) be stored in the `body`, alongside a pre-formatted version in the `formatted_body`

The client is expected to import the `body`, although maybe I should make that more explicit in the document

(or, alternatively to the `body`, if available the `m.draft_full_event` if it's something like a poll or other more complex event, but I have not gotten much feedback on that yet)
